feat: add new contract to support merkl reward claiming

aa-eyup opened the pull request:
## Motivation
Currently there is exists no way for an ATokenVault supplying to Aave to claim Merkl rewards. This change adds a contract that inherits from the upgradeable `ATokenVault` and adds functionality to claim Merkl rewards from a distributor address. If existing ATokenVault instances want to upgrade to support these functions then it can be done as this change consumes one of the gap slots reserved for upgrades to hold the Merkl distributor address.

## Changes
- inherits from ATokenVault
- updates storage struct to store merkl distributor contract (consumes the first of 50 gap slots)
- expose (onlyOwner) function to set distributor
- expose function to get distributor
- expose (onlyOwner) function to claim rewards from merkl distributor
- add test that forks Ethereum mainnet to claim rewards (uses an address that was eligible to claim rewards as of the fork block)

aa-eyup committed:
feat: add new contract to support merkl reward claiming

- inherits from ATokenVault
- updates storage struct to store merkl distributor contract
  (consumes the first of 50 gap slots)
- expose functions to set/get distributor (onlyOwner)
- expose function to claim rewards from merkl distributor (onlyOwner)
- add test that forks Ethereum mainnet to claim rewards

miguelmtzinf commented on src/ATokenVaultMerklRewardClaimer.sol:
```diff
@@ -0,0 +1,75 @@
+import {ATokenVault} from "./ATokenVault.sol";
+import {IATokenVaultMerklRewardClaimer} from "./interfaces/IATokenVaultMerklRewardClaimer.sol";
+
+interface IMerklDistributor {
```
let's move it to a new dir `src/dependencies/merkl/` and leave [reference](https://github.com/AngleProtocol/merkl-contracts/blob/b7bd0e65a3f366e4041bc83494cbd981f8852b16/contracts/Distributor.sol#L195-L206) in the interface. (example: https://github.com/aave/aave-v4/blob/main/src/dependencies/chainlink/AggregatorV3Interface.sol)

miguelmtzinf commented on src/ATokenVaultMerklRewardClaimer.sol:
```diff
@@ -0,0 +1,75 @@
+/**
+ * @title ATokenVaultMerklRewardClaimer
+ * @author Aave Protocol
+ * @notice A contract that allows the owner to claim Merkl rewards for the ATokenVault
```
```suggestion
 * @notice ATokenVault, with Merkl reward claiming capability
```

miguelmtzinf commented on src/ATokenVaultMerklRewardClaimer.sol:
```diff
@@ -0,0 +1,75 @@
+    {}
+
+    /// @inheritdoc IATokenVaultMerklRewardClaimer
+    function getMerklDistributor() external view override returns (address) {
```
write functions must be placed before read functions, per [Solidity style guidelines](https://docs.soliditylang.org/en/latest/style-guide.html#order-of-layout)

miguelmtzinf commented on src/ATokenVaultMerklRewardClaimer.sol:
```diff
@@ -0,0 +1,75 @@
+
+        // The claim function does not return a list of tokens and amounts actually received.
+        // It is possible for rewards to be in aTokens, the underlying asset or some other token.
+        // If necessary the owner can use IATokenVault.emergencyRescue(...) to rescue the non-aToken rewards.
```
we cannot rescue native token and ATokens

miguelmtzinf commented on src/ATokenVaultStorage.sol:
```diff
@@ -20,8 +20,10 @@ abstract contract ATokenVaultStorage {
         uint40 __deprecated_gap;
         // as a fraction of 1e18
         uint64 fee;
+        // Merkl distributor contract address called to claim Merkl rewards
```
```suggestion
        // Merkl distributor contract
```

aa-eyup committed:
enh: address PR comments

- make comments more concise
- clarify in comment that native asset can not be rescued
- move custom interface into a /dependencies dir
- order read/write functions on ATokenVaultMerklRewardsClaim contract and interface

aa-eyup committed:
test: add unit tests for ATokenVaultMerklRewardClaimer

aa-eyup committed:
feat: add function to merkl reward claimer vault to set operator

- an operator will be set on the reward distributor contract
- an operator can claim rewards on behalf of a user
  (rewards are still sent to the user i.e. the ATokenVault)

aa-eyup committed:
enh: include the merkl distributor address in events

aa-eyup committed:
chore: rename test contract

aa-eyup committed:
test: add check for non owner call to toggle operator

miguelmtzinf commented on src/ATokenVaultMerklRewardClaimer.sol:
```diff
@@ -0,0 +1,75 @@
+        // Do not attempt to accrue yield as it can be delegated to subsequent calls to this contract.
+        // We do not need to accrue before claiming because new shares are not granted anywhere (rewards are socialized across all current share holders).
+        // We do not need to accrue after claiming because any subsequent call will trigger an accrual before state updates
+        // and preview functions read the balance of aTokens on the vault at runtime.
```
```suggestion
            users[i] = address(this);
        }

        IMerklDistributor(_s.merklDistributor).claim(users, rewardTokens, amounts, proofs);
```
I think it's completely fine to clean this up

miguelmtzinf commented on src/ATokenVaultMerklRewardClaimer.sol:
```diff
@@ -0,0 +1,75 @@
+        override
+        onlyOwner
+    {
+        require(_s.merklDistributor != address(0), "MERKL_DISTRIBUTOR_NOT_SET");
```
we may want to check `rewardTokens.length == amounts.length == proofs.length`

aa-eyup replied:
Didn't add this because the distributor already validates, but we can check early to avoid extra gas consumption in case of error.

miguelmtzinf commented on src/interfaces/IATokenVaultMerklRewardClaimer.sol:
```diff
@@ -0,0 +1,69 @@
+     * @dev The order of the tokens, amounts, and proofs must align with eachother
+     * @param rewardTokens Addresses of the ERC-20 reward tokens to claim (the tokens passed as params to the Merkl distributor contract)
+     * @param amounts Amounts of the reward tokens to claim for each token
+     * @param proofs Merkl proof passed to the Merkl distributor contract
```
```suggestion
     * @notice Claims Merkl protocol rewards accrued from vault deposits.
     * @dev Only callable by the owner
     * @dev The order of elements in the rewardTokens, amounts, and proofs arrays must correspond precisely
     * @dev Claimed tokens are retained within the contract and may subsequently be withdrawn using the emergencyRescue function
     * @param rewardTokens The array of ERC-20 token addresses representing the reward tokens to be claimed
     * @param amounts The array specifying the amount to claim for each corresponding reward token
     * @param proofs The array containing the merkle proofs required for each reward claim
```

miguelmtzinf commented on src/interfaces/IATokenVaultMerklRewardClaimer.sol:
```diff
@@ -0,0 +1,69 @@
+
+    /**
+     * @notice Getter for the contract address called to claim Merkl rewards
+     * @return Address of the Merkl distributor contract
```
```suggestion
     * @notice Returns the address of the Merkl distributor contract.
     * @return The address of the Merkl distributor contract
```

miguelmtzinf commented on src/dependencies/merkl/DistributorInterface.sol:
```diff
@@ -0,0 +1,16 @@
+// Based on implementation from https://github.com/AngleProtocol/merkl-contracts/blob/b7bd0e65a3f366e4041bc83494cbd981f8852b16/contracts/Distributor.sol#L202
+// All Rights Reserved © AaveCo
+
+pragma solidity ^0.8.10;
```
```suggestion
// SPDX-License-Identifier: BUSL-1.1
// Based on implementation from https://github.com/AngleProtocol/merkl-contracts/blob/b7bd0e65a3f366e4041bc83494cbd981f8852b16/contracts/Distributor.sol#L202
pragma solidity ^0.8.10;
```

miguelmtzinf commented on test/ATokenVaultMerklRewardClaimer.t.sol:
```diff
@@ -0,0 +1,201 @@
+    MockAavePool pool;
+    MockAToken aDai;
+    MockDAI dai;
+    IATokenVaultMerklRewardClaimer vaultMerklRewardClaimer;
```
add internal modifier for clarity

miguelmtzinf commented on test/ATokenVaultMerklRewardClaimer.t.sol:
```diff
@@ -0,0 +1,201 @@
+        assertEq(merklDistributor.getLastAmounts().length, 1);
+        assertEq(merklDistributor.getLastAmounts()[0], 1000);
+        assertEq(merklDistributor.getLastProofs().length, 1);
+        assertEq(merklDistributor.getLastProofs()[0][0], proof);
```
we can use `expectCall` instead. [docs](https://getfoundry.sh/reference/cheatcodes/expect-call#description)
We can rm the part where we store the call in the Mock. We could remove the mock entirely too I think, and just prank the merkl distributor to transfer assets to the vault

aa-eyup replied:
I do not think we can trigger side effects from addresses that are called unless we use a mock. You're right `expectCall` is suitable for this test. For example mocking a revert or a token transfer once a function is called on a target I think requires a mock contract.

miguelmtzinf commented on test/ATokenVaultMerklRewardClaimer.t.sol:
```diff
@@ -0,0 +1,201 @@
+        (address[] memory rewardTokens, uint256[] memory amounts, bytes32[][] memory proofs) = _buildMerklRewardsClaimData(address(dai), 1000, proof);
+        vm.prank(OWNER);
+        vm.expectEmit(true, true, false, true, address(vaultMerklRewardClaimer));
+        emit IATokenVaultMerklRewardClaimer.MerklRewardsClaimed(address(merklDistributor), rewardTokens, amounts);
```
event claim can be check in the same test case for a successfull claim call

aa-eyup replied:
I agree - I was just staying consistent with how other tests in this repo were written. Will make a change.

miguelmtzinf commented on test/ATokenVaultMerklRewardClaimer.t.sol:
```diff
@@ -0,0 +1,201 @@
+
+    function testClaimMerklRewardsRevertsIfMerklDistributorNotSet() public {
+        bytes32 proof = keccak256("proof1");
+        (address[] memory rewardTokens, uint256[] memory amounts, bytes32[][] memory proofs) = _buildMerklRewardsClaimData(address(dai), 1000, proof);
```
not needed

miguelmtzinf commented on test/ATokenVaultMerklRewardClaimer.t.sol:
```diff
@@ -0,0 +1,201 @@
+
+    function testClaimMerklRewardsRevertsIfNotOwner() public {
+        bytes32 proof = keccak256("proof1");
+        (address[] memory rewardTokens, uint256[] memory amounts, bytes32[][] memory proofs) = _buildMerklRewardsClaimData(address(dai), 1000, proof);
```
same

miguelmtzinf commented on test/ATokenVaultMerklRewardClaimer.t.sol:
```diff
@@ -0,0 +1,201 @@
+        bytes32 proof = keccak256("proof1");
+        (address[] memory rewardTokens, uint256[] memory amounts, bytes32[][] memory proofs) = _buildMerklRewardsClaimData(address(dai), 1000, proof);
+        string memory revertReason = "revert because of insufficient balance";
+        merklDistributor.setShouldRevert(true, revertReason);
```
ok, we need the mock for this

miguelmtzinf commented on test/ATokenVaultMerklRewardClaimer.t.sol:
```diff
@@ -0,0 +1,201 @@
+        assertEq(vaultMerklRewardClaimer.getMerklDistributor(), address(merklDistributor));
+    }
+
+    function testSetMerklDistributorEmitsEvent() public {
```
merge pls

aa-eyup replied:
addressed

miguelmtzinf commented on test/ATokenVaultMerklRewardClaimer.t.sol:
can we have tests for claiming aTokens or native tokens please?
can we have tests for claiming and rescuing?
can we have a test scenario where we claim aToken and gets eaten by suppliers interest?

aa-eyup replied:
added these test cases

aa-eyup committed:
refactor(ATokenVaultMerklRewardClaimer): address feedback

miguelmtzinf commented on src/ATokenVaultMerklRewardClaimer.sol:
```diff
@@ -0,0 +1,58 @@
+
+import {ATokenVault} from "./ATokenVault.sol";
+import {IATokenVaultMerklRewardClaimer} from "./interfaces/IATokenVaultMerklRewardClaimer.sol";
+import {IMerklDistributor} from "./dependencies/merkl/DistributorInterface.sol";
```
```suggestion
import {IPoolAddressesProvider} from "@aave-v3-core/interfaces/IPoolAddressesProvider.sol";
import {IERC20} from "@openzeppelin/interfaces/IERC20.sol";
import {IMerklDistributor} from "./dependencies/merkl/DistributorInterface.sol";
import {ATokenVault} from "./ATokenVault.sol";
import {IATokenVaultMerklRewardClaimer} from "./interfaces/IATokenVaultMerklRewardClaimer.sol";
```

miguelmtzinf commented on src/interfaces/IATokenVaultMerklRewardClaimer.sol:
```diff
@@ -0,0 +1,55 @@
+/**
+ * @title IATokenVaultMerklRewardClaimer
+ * @author Aave Protocol
+ *
+ * @notice Defines the basic interface of the ATokenVaultMerklRewardClaimer
```
```suggestion
 * @author Aave Protocol
 * @notice Defines the basic interface of the ATokenVaultMerklRewardClaimer
```

miguelmtzinf commented on src/ATokenVaultStorage.sol:
```diff
@@ -20,8 +20,10 @@ abstract contract ATokenVaultStorage {
+        address merklDistributor;
         // Reserved storage space to allow for layout changes in the future
-        uint256[50] __gap;
+        uint256[49] __gap;
```
hmm not sure this makes sense if we don't aim to do any upgrade

aa-eyup replied:
Should we not allow existing deployments to upgrade to a version that support merkl reward claiming?

miguelmtzinf commented on test/mocks/MockMerklDistributor.sol:
```diff
@@ -0,0 +1,55 @@
+        }
+        _recipients = new address[](0);
+        _tokens = new address[](0);
+        _amounts = new uint256[](0);
```
why do we need this?

aa-eyup replied:
We actually don't since a claim is only called once per test. I initially had it to allow clearing mock state within a test. Will remove.

miguelmtzinf commented on test/ATokenVaultMerklRewardClaimer.t.sol:
```diff
@@ -0,0 +1,306 @@
+        // Avoid stack too deep error
+        _claimMerklRewards();
+
+        uint256 vaultATokenBalanceAfter = _aDai.balanceOf(address(vault));
```
do we wanna add a call to emergencyRescue(underlyinAToken) and expect revert? just to showcase

aa-eyup replied:
Added.

miguelmtzinf commented on test/ATokenVaultMerklRewardClaimer.t.sol:
```diff
@@ -0,0 +1,306 @@
+        vm.deal(address(_merklDistributor), amountOfNativeToken);
+        
+        bytes32 proof = keccak256("proof1");
+        (address[] memory rewardTokens, uint256[] memory amounts, bytes32[][] memory proofs) = _buildMerklRewardsClaimData(address(_dai), 1000, proof);
```
why `_dai`? if this is the test case for native?

aa-eyup replied:
The thinking was that you can input a token to claim, but receive a different asset.

miguelmtzinf replied:
i think it's clearer to use weth since this case is for receiving native token. I d add another test case for receiveing something diff than input.

aa-eyup replied:
>i think it's clearer to use weth since this case is for receiving native token. 

Created a `MockWETH` token and used it in this test.

> I d add another test case for receiveing something diff than input.

We already have a test for receiving something different than input (`testClaimMerklRewardsIfATokenIsRewarded` where input is `dai` and reward is `_aDai`).

miguelmtzinf commented on test/ATokenVaultMerklRewardClaimerFork.t.sol:
```diff
@@ -0,0 +1,204 @@
+
+    function testEthereumForkBalanceOfAddressWithClaimableRewards() public {
+        assertEq(ADDRESS_WITH_CLAIMABLE_REWARDS.balance, 154288817306978598);
+    }
```
i think first 2 are not needed, and the 3rd can be moved to the test case in particular.

miguelmtzinf replied:
Also, let's not duplicate test cases. Imho it's fine to remove all unit tests that can be found in the other file, and focus here on the integration with Merkl.

aa-eyup replied:
> i think first 2 are not needed
Ack - again, was just being consistent with the other test files. Will remove the first two. Will also remove the 3rd as it was just a sanity check that the forked state is correct.

Also I removed all unit test-like tests in this file and just kept the actual integration test that relies on the forked state.

aa-eyup committed:
enh(ATokenVaultMerklRewardClaimer): address feedback on additional tests and cleanup

aa-eyup committed:
feat(ATokenVaultMerklRewardClaimer): allow setting destination to forward reward tokens to

miguelmtzinf commented on src/interfaces/IATokenVaultMerklRewardClaimer.sol:
```diff
@@ -20,6 +20,14 @@ interface IATokenVaultMerklRewardClaimer {
+     * @param destination Address of the destination receiving the reward token
+     * @param amount Amount of the reward token transferred to the destination
+     */
+    event MerklRewardsTokenForwarded(address indexed token, address indexed destination, uint256 indexed amount);
```
We could do smt similar to `RewardsClaimed`, and emit the destination, tokens array and amounts array.

aa-eyup replied:
I have no strong preference, but I wanted to avoid emitting the event if the `rewardTokensToForward` array is empty. If you have a strong preference please lmk. Maybe we return early if `rewardTokensToForward` is empty then iterate through the tokens then emit an event at the event.

Imagine it would look like this:
`event MerklRewardsTokenForwarded(address indexed destination, address[] tokens, uint256[] amounts);`

We can also pass in an array of destinations to `claimMerklRewards(...)` if you think the flexibility would be valuable - then the single token/amount/destination event makes more sense.

miguelmtzinf commented on test/mocks/MockMerklDistributor.sol:
it's always the caller, but fine to keep it as is

aa-eyup committed:
test(ATokenVaultMerklRewardClaimer): use weth mock token to test native reward claim

aa-eyup committed:
chore(IATokenVaultMerklRewardClaimer): format docs

donosonaumczuk committed:
Merge branch 'main' into feat/merkl-rewards

donosonaumczuk committed:
misc: ATokenVaultMerklRewardClaimer deployment script

aa-eyup committed:
fix(ATokenVaultMerklRewardClaimer): do not allow claiming and forwarding of aToken

- The change is intended to protect against an attack which leverages a
  malicious distributor contract which enters into the ATokenVault
  via depositATokens(uint256 assets, address receiver) then forwards
  the aToken after the IMerklDistributor::claim call executes,
  effectively allowing an actor to obtain vault shares for free
